### PR TITLE
Add JSON Configuration plugin.

### PR DESCRIPTION
**Pull request by marcusramberg**
Support filename.mode
file can be relative to home or absolute path
Also adds existsSync

### CODE REVIEW

File: test/jsonconfig-app.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import t from 'tap';
+import File from '../lib/file.js';
+import {app} from './support/app-config/app-config.js';
+
+t.test('JSONConfig app', async t => {
+  const client = await app.newTestClient({tap: t});
+
+  await t.test('External app with mode file', async t => {
+    t.same(app.config, {name: 'Bond. James Bond', drink: 'Martini', car: 'Aston Martin'});
+    (await client.getOk('/')).statusIs(200).bodyIs('My name is Bond. James Bond');
```
Comment (kraih) [inline test/jsonconfig-app.js:12]: Missing punctuation?

File: lib/plugins/json-config.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import File from '../file.js';
+import path from 'path';
```
Comment (kraih) [inline lib/plugins/json-config.js:2]: Why do you still need to import `path`?

File: lib/plugins/json-config.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+export default function jsonConfigPlugin (app, options) {
```
Comment (kraih) [inline lib/plugins/json-config.js:3]: Why no whitespace?

File: lib/plugins/json-config.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  const filename = options?.file ?? 'config.json';
+  const filePath = path.isAbsolute(filename) ? new File(filename) : app.home.child(filename);
+  const fileParts = filePath.toString().split('.');
+  fileParts.splice(-1, 0, app.mode);
+  const modeFilePath = new File(fileParts.join('.'));
+  let modeConfig = {};
```
Comment (kraih) [inline lib/plugins/json-config.js:9]: I don't think assigning an empty object does anything.

File: lib/plugins/json-config.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  if (modeFilePath.existsSync()) {
+    modeConfig = JSON.parse(modeFilePath.readFileSync().toString());
+  }
```
Comment (kraih) [inline lib/plugins/json-config.js:12]: This could be one line.

File: lib/plugins/json-config.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  const config = JSON.parse(filePath.readFileSync());
+  app.config = {...app.config, ...config, ...modeConfig};
```
Comment (kraih) [inline lib/plugins/json-config.js:14]: You could change the order and just use `Object.assign()` to make the code cleaner.

File: package.json [hunk 7]
```diff
@@ -38,6 +39,7 @@
     "cookie": ">=0.4.1",
     "ejs": ">=3.1.6",
     "escape-string-regexp": ">=2.0.0",
+    "eslint-config-standard": "^16.0.2",
```
Comment (kraih) [inline package.json:42]: This looks like an accident.

File: package.json [hunk 8]
```diff
@@ -26,6 +26,7 @@
   "type": "module",
   "devDependencies": {
     "@istanbuljs/esm-loader-hook": "^0.1.2",
+    "@types/tap": "^15.0.1",
```
Comment (kraih) [inline package.json:29]: Why is the version pinned?

File: test/jsonconfig-app.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
+import mojo from '../lib/mojo.js';
+import jsonConfigPlugin from '../lib/plugins/json-config.js';
+import t from 'tap';
+import File from '../lib/file.js';
+import {app} from './support/app-config/app-config.js';
```
Comment (kraih) [inline test/jsonconfig-app.js:5]: Not ordered alphabetically.

File: test/jsonconfig-app.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    (await client.getOk('/')).statusIs(200).bodyIs('My name is Bond. James Bond');
+    t.end();
+  });
+  await client.stop();
+
+  await t.test('Absolute config file path', async t => {
+    const app = mojo();
+    app.plugin(jsonConfigPlugin, {file: File.currentFile().dirname().child('support', 'named.json').toString()});
+    t.same(app.config, {name: 'namedConfig', deep: {option: ['deep']}});
+  });
```
Comment (kraih) [inline test/jsonconfig-app.js:21]: This subtest is not actually async.

File: test/jsonconfig-app.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+  await t.test('Missing config file', async t => {
+    const app = mojo();
+    t.throws(function () {
+      app.plugin(jsonConfigPlugin(app, {file: 'missing.json'}));
+    }, 'ENOENT: no such file or directory');
+  });
```
Comment (kraih) [inline test/jsonconfig-app.js:28]: Also not actually async.

File: test/support/app-config/app-config.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import mojo from '../../../lib/mojo.js';
+import jsonConfigPlugin from '../../../lib/plugins/json-config.js';
```
Comment (kraih) [inline test/support/app-config/app-config.js:2]: Nor ordered alphabetically.

File: test/support/app-config/config.json [hunk 13]
```diff
@@ -0,0 +1 @@
+{"name":"Bond. James Bond"}
```
Comment (kraih) [inline test/support/app-config/config.json:1]: Is the missing newline at the end part of the test case?

**Review by kraih — CHANGES_REQUESTED** [review]

**Comment by kraih** [comment]
Please use proper grammar in the commit message. I would have also preferred a separate PR for `existsSync`.

File: lib/plugins/json-config.js [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+import path from 'path';
+export default function jsonConfigPlugin (app, options) {
+  const filename = options?.file ?? 'config.json';
+  const filePath = path.isAbsolute(filename) ? new File(filename) : app.home.child(filename);
+  const fileParts = filePath.toString().split('.');
+  fileParts.splice(-1, 0, app.mode);
+  const modeFilePath = new File(fileParts.join('.'));
+  let modeConfig = {};
+  if (modeFilePath.existsSync()) {
+    modeConfig = JSON.parse(modeFilePath.readFileSync().toString());
```
Comment (kraih) [inline lib/plugins/json-config.js:11]: The `toString()` looks like it does nothing.

**Comment by kraih** [comment]
Looks like you're also not testing the load order with an override at each level.

File: lib/plugins/json-config.js [hunk 15]
```diff
@@ -0,0 +1,15 @@
+import File from '../file.js';
+import path from 'path';
+export default function jsonConfigPlugin (app, options) {
+  const filename = options?.file ?? 'config.json';
+  const filePath = path.isAbsolute(filename) ? new File(filename) : app.home.child(filename);
+  const fileParts = filePath.toString().split('.');
+  fileParts.splice(-1, 0, app.mode);
+  const modeFilePath = new File(fileParts.join('.'));
+  let modeConfig = {};
+  if (modeFilePath.existsSync()) {
```
Comment (kraih) [inline lib/plugins/json-config.js:10]: I also prefer type checks whenever possible, so `=== true`.

File: package.json (hunk 8 above)
Comment (kraih) [reply package.json:29]: I guess this also doesn't really belong i this PR.

File: test/jsonconfig-app.js [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+  await t.test('Absolute config file path', async t => {
+    const app = mojo();
+    app.plugin(jsonConfigPlugin, {file: File.currentFile().dirname().child('support', 'named.json').toString()});
+    t.same(app.config, {name: 'namedConfig', deep: {option: ['deep']}});
+  });
+
+  await t.test('Missing config file', async t => {
+    const app = mojo();
+    t.throws(function () {
```
Comment (kraih) [inline test/jsonconfig-app.js:25]: This could probably be an arrow function.

File: package.json (hunk 7 above)
Comment (marcusramberg) [reply package.json:42]: Eslint didn't work for me without this dependency. 🤷

Comment (kraih) [reply package.json:42]: Still doesn't belong in the PR, and/or non dev dependencies.

File: lib/plugins/json-config.js [hunk 17]
```diff
@@ -0,0 +1,12 @@
...
+
+export default function jsonConfigPlugin (app, options) {
+  const filename = new File(options?.file ?? 'config.json');
+  const filePath = filename.isAbsolute() ? filename : app.home.child(filename.toString());
+  const fileParts = filePath.toString().split('.');
+  fileParts.splice(-1, 0, app.mode);
+  const modeFilePath = new File(fileParts.join('.'));
+  let modeConfig;
+  if (modeFilePath.existsSync() === true) modeConfig = JSON.parse(modeFilePath.readFileSync());
+  Object.assign(app.config, JSON.parse(filePath.readFileSync()), modeConfig);
```
Comment (kraih) [inline lib/plugins/json-config.js:11]: Use two `Object.assign` calls and you can get rid of the `let modeConfig`.

File: test/jsonconfig-app.js [hunk 18]
```diff
@@ -0,0 +1,31 @@
+import {app} from './support/app-config/app-config.js';
+import File from '../lib/file.js';
+import mojo from '../lib/mojo.js';
+import jsonConfigPlugin from '../lib/plugins/json-config.js';
```
Comment (kraih) [inline test/jsonconfig-app.js:4]: Alphabetical order.

File: test/support/app-config/app-config.js [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+import mojo from '../../../lib/mojo.js';
+
+export const app = mojo();
+
+app.config = {drink: 'Martini'};
+app.plugin(jsonConfigPlugin);
+
+app.any('/', ctx => ctx.render({text: 'My name is ' + app.config.name + '.'}));
```
Comment (kraih) [inline test/support/app-config/app-config.js:9]: I'd probably use a template expression there:
```
`My name is ${app.config.name}.`
```

File: test/jsonconfig-app.js [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+    (await client.getOk('/')).statusIs(200).bodyIs('My name is Bond. James Bond.');
+  });
+  await client.stop();
+
+  t.test('Default JSONConfig App', async t => {
+    const app = mojo();
+    app.config = {name: 'overridden', extra: 'option'};
+    app.plugin(jsonConfigPlugin, {file: File.currentFile().dirname().child('support', 'named.json').toString()});
+    t.same(app.config, {name: 'namedConfig', extra: 'option', deep: {option: ['deep']}});
+    t.end();
```
Comment (kraih) [inline test/jsonconfig-app.js:21]: I think you're missing a test case here where you change the mode of the app and do not find a mode specific config and a value is not redefined from the main config.

**Pull request merged**